src/slam_karto.cpp

```diff
@@ -291,12 +345,19 @@ SlamKarto::~SlamKarto()
     transform_thread_->join();
     delete transform_thread_;
   }
+  if(vis_thread_)
+  {
+    vis_thread_->join();
+    delete vis_thread_;
+  }
+  if(solver_)
```

Comment:
I suspect the problem comes from the fact that the `SPAGraphVisualizer` holds a `shared_ptr` to `solver_`. If I comment out all the code of the visualizer in slam_karto and leave only the solver plugin, the warning disappears.

If `reset()` does not effectively delete the objects, I guess it means some other object having shared ownership over the pointer has not yet been deleted and the reference counter is still > 1.
The 2 objects sharing ownership of the `solver_` with the slam_karto node are the `karto::Mapper` pointed to by `mapper_` and the `karto::GraphVisualizer` pointed to by `visualizer_`.  However even if I make sure that they get deleted/reset before `solver_` in `~SlamKarto()`, I still get the same warning.
So far the only thing that worked was to delete the raw pointer with `delete smart_ptr.get()`. Which does look bad.

An alternative might be to use a different smart pointer with single ownership for the slam karto node.